xds/src/main/java/io/grpc/xds/XdsClientWrapperForServerSds.java

```diff
@@ -170,31 +173,198 @@ public void onError(Status error) {
+      } else if (sourceType == EnvoyServerProtoData.ConnectionSourceType.EXTERNAL) {
+        matching = !sourceAddress.isLoopbackAddress() && !sourceAddress.isAnyLocalAddress();
+      } else { // ANY or null
+        matching = true;
+      }
+      if (matching) {
+        filtered.add(filterChain);
+      }
+    }
+    return filtered;
+  }
+
+  private static boolean isCidrMatching(byte[] cidrBytes, byte[] addressBytes, int prefixLen) {
+    BigInteger cidrInt = new BigInteger(cidrBytes);
+    BigInteger addrInt = new BigInteger(addressBytes);
+
+    int shiftAmount = 8 * cidrBytes.length - prefixLen;
+
+    cidrInt = cidrInt.shiftRight(shiftAmount);
+    addrInt = addrInt.shiftRight(shiftAmount);
+    return cidrInt.equals(addrInt);
+  }
+
+  private static class QueueElement {
+    FilterChain filterChain;
+    int indexOfMatchingPrefixRange;
+    int matchingPrefixLength;
+
+    public QueueElement(FilterChain filterChain, InetAddress address, boolean forDestination) {
+      this.filterChain = filterChain;
+      FilterChainMatch filterChainMatch = filterChain.getFilterChainMatch();
+      byte[] addressBytes = address.getAddress();
+      boolean isIPv6 = address instanceof Inet6Address;
+      List<CidrRange> cidrRanges =
+          forDestination
+              ? filterChainMatch.getPrefixRanges()
+              : filterChainMatch.getSourcePrefixRanges();
+      indexOfMatchingPrefixRange = -1;
+      if (cidrRanges.isEmpty()) { // if there is no CidrRange assume there is perfect match
+        matchingPrefixLength = isIPv6 ? 128 : 32;
+      } else {
+        matchingPrefixLength = 0;
+        int index = 0;
+        for (CidrRange cidrRange : cidrRanges) {
+          InetAddress cidrAddr = cidrRange.getAddressPrefix();
+          boolean cidrIsIpv6 = cidrAddr instanceof Inet6Address;
+          if (isIPv6 == cidrIsIpv6) {
+            byte[] cidrBytes = cidrAddr.getAddress();
+            int prefixLen = cidrRange.getPrefixLen();
+            if (isCidrMatching(cidrBytes, addressBytes, prefixLen)
+                && prefixLen > matchingPrefixLength) {
+              matchingPrefixLength = prefixLen;
+              indexOfMatchingPrefixRange = index;
+            }
+          }
+          index++;
+        }
+      }
+    }
+  }
+
+  private static final class QueueElementComparator implements Comparator<QueueElement> {
+
+    @Override
+    public int compare(QueueElement o1, QueueElement o2) {
+      // descending order for max heap
+      return o2.matchingPrefixLength - o1.matchingPrefixLength;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      return obj instanceof QueueElementComparator;
+    }
+
+    @Override
+    public int hashCode() {
+      return super.hashCode();
+    }
+  }
+
+  // use prefix_ranges (CIDR) and get the most specific matches
+  private List<FilterChain> filterOnIpAddress(
+      List<FilterChain> filterChains, InetAddress address, boolean forDestination) {
```

Comment:
The comment on the `prefix_ranges` field says the following:
```
	// If non-empty, an IP address and prefix length to match addresses when the
	// listener is bound to 0.0.0.0/:: or when use_original_dst is specified.
```
Are you checking if the listener is bound to `0.0.0.0/::` here?

Reply:
In [grpc-java it is always bound](https://github.com/grpc/grpc-java/blob/master/xds/src/main/java/io/grpc/xds/XdsClientWrapperForServerSds.java#L129) to `0.0.0.0` .